### PR TITLE
feat: mocks by query & body + object syntax

### PR DESCRIPTION
**Pull request by AndersDJohnson**
Adds ability to define mocks for specific combinations of query parameters and request bodies. Requests will only match if at least the defined query parameters are passed, or if the request body matches. Query parameters can be defined directly in the path for the mock, i.e., `mockyeah.get('/my/path?key=value&plus=more')`, whereas specifying a body requires a new object syntax described as follows. The object syntax can be used in place of the URL path, because sometimes that API might be preferred for providing query parameters, and also will enable future matching features (e.g., headers, functions, regular expressions, etc.):
```
mockyeah.post({
  path: '/my/path',
  query: {
    key: 'value',
    plus: 'more'
  },
  body: {
    something: 'here'
  }
})
```

This uses Node's `url.parse` query parameter parsing logic (and whatever Express uses - hopefully similar enough). So it handles repeated query parameters as arrays, so `?a=1&a=2` would be:
```
{
  a: ['1', '2']
}
```

BREAKING CHANGE: Previously, query parameters on mock definitions were ignored. So mocks defined uselessly with query parameters would match requests for the same path even without the defined query parameters being passed. I opted to break backwards compatibility here because it seems like the right behavior going forward, rather than requiring an additional option to be passed or requiring the object syntax for new behavior.

### CODE REVIEW

File: app/lib/RouteResolver.js [hunk 1]
```diff
@@ -92,7 +109,17 @@ function RouteResolver(app) {
...
-  route.pathname = normalizePathname(parse(route.path, true).pathname);
+  if (typeof path === 'string') {
+    const url = parse(route.path, true);
+    route.pathname = normalizePathname(url.pathname);
+    route.query = url.query;
+  } else {
+    const object = route.path;
+    route.pathname = normalizePathname(object.path);
+    route.query = object.query || null; // because `url.parse` returns `null`
+    route.body = object.body;
```
Comment (ryanricard) [inline app/lib/RouteResolver.js]: `pathObject` > `object` may improve readability here.

File: app/lib/RouteResolver.js [hunk 2]
```diff
@@ -33,13 +33,30 @@ function isRouteForRequest(route, req) {
 
   if (route.pathname !== '*' && !route.pathRegExp.test(pathname)) return false;
 
-  // TODO: Later add features to match other things, like query parameters, etc.
+  const matchesParams = _.every(route.query, (value, key) => {
+    return _.isEqual(_.get(req.query, key), value);
```
Comment (ryanricard) [inline app/lib/RouteResolver.js]: I am fine with destructuring `lodash` methods. e.g. `const { get } = require('lodash')`.

File: app/lib/RouteResolver.js [hunk 3]
```diff
@@ -33,13 +33,30 @@ function isRouteForRequest(route, req) {
 
   if (route.pathname !== '*' && !route.pathRegExp.test(pathname)) return false;
 
-  // TODO: Later add features to match other things, like query parameters, etc.
+  const matchesParams = _.every(route.query, (value, key) => {
```
Comment (ryanricard) [inline app/lib/RouteResolver.js]: I assume you chose `_.every` because `route.query` is an object.

Comment (AndersDJohnson) [reply app/lib/RouteResolver.js]: Correct.

File: app/lib/RouteResolver.js (hunk 2 above)
Comment (AndersDJohnson) [reply app/lib/RouteResolver.js]: I can refactor. This PR or another?

File: app/lib/RouteResolver.js (hunk 1 above)
Comment (AndersDJohnson) [reply app/lib/RouteResolver.js]: Will refactor.

**Pull request merged**